firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java

```diff
@@ -0,0 +1,68 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.local;
+
+import static com.google.firebase.firestore.util.Assert.hardAssert;
+
+import com.google.firebase.firestore.model.ResourcePath;
+import java.util.ArrayList;
+import java.util.List;
+
+/** A persisted implementation of IndexManager. */
+final class SQLiteIndexManager implements IndexManager {
+  /**
+   * An in-memory copy of the index entries we've already written since the SDK launched. Used to
+   * avoid re-writing the same entry repeatedly.
+   *
+   * <p>This is *NOT* a complete cache of what's in persistence and so can never be used to satisfy
+   * reads.
+   */
+  private final MemoryIndexManager.MemoryCollectionParentIndex collectionParentsCache =
+      new MemoryIndexManager.MemoryCollectionParentIndex();
+
+  private final SQLitePersistence db;
+
+  SQLiteIndexManager(SQLitePersistence persistence) {
+    db = persistence;
+  }
+
+  @Override
+  public void addToCollectionParentIndex(ResourcePath collectionPath) {
+    hardAssert(collectionPath.length() % 2 == 1, "Expected a collection path.");
+    if (collectionParentsCache.add(collectionPath)) {
+      hardAssert(collectionPath.length() >= 1, "Invalid collection path.");
+      String collectionId = collectionPath.getLastSegment();
```

Comment:
Optional: add a newline before this line?

Reply:
This hardAssert is actually redundant with the first one, so I just removed it.